File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/PolygonsController.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import com.esri.arcgisruntime.mapping.view.Graphic;
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;
import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodChannel;

public class PolygonsController extends BaseSymbolController implements MapTouchGraphicDelegate {
    private final Map<String, PolygonController> polygonIdToController = new HashMap<>();

    private final MethodChannel methodChannel;
    private final GraphicsOverlay graphicsOverlay;
    private final SelectionPropertiesHandler selectionPropertiesHandler;


    public PolygonsController(MethodChannel methodChannel, GraphicsOverlay graphicsOverlay, SelectionPropertiesHandler selectionPropertiesHandler) {
        this.methodChannel = methodChannel;
        this.graphicsOverlay = graphicsOverlay;
        this.selectionPropertiesHandler = selectionPropertiesHandler;
    }

    @Override
    public boolean canConsumeTaps() {
        for (final PolygonController controller : polygonIdToController.values()) {
            if (controller.canConsumeTapEvents()) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean didHandleGraphic(Graphic graphic) {
        final Object rawPolygonId = graphic.getAttributes().get("polygonId");
        if (rawPolygonId == null) {
            return false;
        }
        final String polygonId = (String) rawPolygonId;
        final PolygonController controller = polygonIdToController.get(polygonId);
        if (controller == null || !controller.canConsumeTapEvents()) {
            return false;
        }
        methodChannel.invokeMethod("polygon#onTap", Convert.polygonIdToJson(polygonId));
        return true;
    }

    public void addPolygons(List<Object> polygonsToAdd) {
        if (polygonsToAdd == null) {
            return;
        }

        executor.execute(() -> {
            for (Object polygon : polygonsToAdd) {
                final Map<?, ?> data = (Map<?, ?>) polygon;
                if (data == null) {
                    continue;
                }
                final String polygonId = (String) data.get("polygonId");
                final PolygonController controller = new PolygonController(polygonId, selectionPropertiesHandler);
                polygonIdToController.put(polygonId, controller);
                Convert.interpretPolygonController(data, controller);
                controller.add(graphicsOverlay);
            }
        });
    }

    public void changePolygons(List<Object> polygonsToChange) {
        if (polygonsToChange == null) {
            return;
        }
        executor.execute(() -> {
            for (Object polygon : polygonsToChange) {
                final Map<?, ?> data = (Map<?, ?>) polygon;
                if (data == null) {
                    continue;
                }
                final String polygonId = (String) data.get("polygonId");
                final PolygonController controller = polygonIdToController.get(polygonId);
                if (controller == null) {
                    continue;
                }
                Convert.interpretPolygonController(data, controller);
            }
        });
    }

    public void removePolygons(List<Object> polygonIdsToRemove) {
        if (polygonIdsToRemove == null) {
            return;
        }
        executor.execute(() -> {
            for (Object rawPolygonId : polygonIdsToRemove) {
                if (rawPolygonId == null) {
                    continue;
                }
                final String polygonId = (String) rawPolygonId;
                final PolygonController controller = polygonIdToController.remove(polygonId);
                if (controller != null) {
                    controller.remove(graphicsOverlay);
                }
            }
        });
    }
}

File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/mapping/symbology/PolygonsController.kt
package com.valentingrigorean.arcgis_maps_flutter.mapping.symbology

import com.arcgismaps.mapping.view.Graphic
import com.arcgismaps.mapping.view.GraphicsOverlay
import com.valentingrigorean.arcgis_maps_flutter.convert.map.toPolygonIdValue
import com.valentingrigorean.arcgis_maps_flutter.map.MapTouchGraphicDelegate
import io.flutter.plugin.common.MethodChannel

class PolygonsController(
    private val methodChannel: MethodChannel,
    private val graphicsOverlay: GraphicsOverlay
) : BaseSymbolController(), MapTouchGraphicDelegate {
    private val polygonIdToController: MutableMap<String?, PolygonController> = HashMap()
    override fun canConsumeTaps(): Boolean {
        for (controller in polygonIdToController.values) {
            if (controller.consumeTapEvents) {
                return true
            }
        }
        return false
    }

    override fun didHandleGraphic(graphic: Graphic): Boolean {
        val rawPolygonId = graphic.attributes["polygonId"] ?: return false
        val polygonId = rawPolygonId as String
        val controller = polygonIdToController[polygonId]
        if (controller == null || !controller.consumeTapEvents) {
            return false
        }
        methodChannel.invokeMethod("polygon#onTap", polygonId.toPolygonIdValue())
        return true
    }

    fun addPolygons(polygonsToAdd: List<Any?>?) {
        if (polygonsToAdd == null) {
            return
        }
        for (polygon in polygonsToAdd) {
            val data = polygon as Map<*, *>? ?: continue
            val polygonId = data["polygonId"] as String
            val controller = PolygonController(polygonId)
            controller.setSelectionPropertiesHandler(selectionPropertiesHandler)
            controller.interpretGraphicController(data, symbolVisibilityFilterController)
            polygonIdToController[polygonId] = controller
            controller.add(graphicsOverlay)
        }
    }

    fun changePolygons(polygonsToChange: List<Any?>?) {
        if (polygonsToChange == null) {
            return
        }
        for (polygon in polygonsToChange) {
            val data = polygon as Map<*, *>? ?: continue
            val polygonId = data["polygonId"] as String?
            val controller = polygonIdToController[polygonId] ?: continue
            controller.interpretGraphicController(data, symbolVisibilityFilterController)
        }
    }

    fun removePolygons(polygonIdsToRemove: List<Any?>?) {
        if (polygonIdsToRemove == null) {
            return
        }
        for (rawPolygonId in polygonIdsToRemove) {
            if (rawPolygonId == null) {
                continue
            }
            val polygonId = rawPolygonId as String
            val controller = polygonIdToController.remove(polygonId)
            if (controller != null) {
                onSymbolRemoval(controller)
                controller.remove(graphicsOverlay)
            }
        }
    }
}